azavea/echo-locator
# Add translation support to dock dropdown menus

## Overview

This PR adds translation support to the 'from' and 'network' dropdown menus in the dock. While translation support was added in [PR 377](https://github.com/azavea/echo-locator/pull/377) and translations strings were added in for most of the  the website in [PR 390](https://github.com/azavea/echo-locator/pull/390), translating the menus was more complex due to the connection between those menus and state. 


### Demo

![Screen Shot 2022-02-01 at 8 13 06 AM](https://user-images.githubusercontent.com/77936689/151975008-661f6327-fc28-4d64-8407-998d8ada4a10.png)

![Screen Shot 2022-02-01 at 8 13 16 AM](https://user-images.githubusercontent.com/77936689/151975029-bd4b9bd9-298e-4d09-814f-9d815e694c67.png)

The 'from' dropdown menu translated into both Chinese and Spanish.

![Screen Shot 2022-02-01 at 8 13 28 AM](https://user-images.githubusercontent.com/77936689/151975013-1d37203c-b842-4a48-aedf-89ce6e8ea181.png)

The 'network' menu translated into Chinese. There are currently only 2 translated options for this menu; placeholder English strings have been used for all 4 options in Spanish and the "Peak No Express" and "Off Peak No Express" options in Chinese.


### Notes

**Consolidation**
Initially, the addition of the `'TripPurpose.'` prefix used for `react-18next` to reference the relevant translation was being done in 3 different spots: `neighborhood-detail.js`, `route-card.js`, and `edit-profile.js`, with state storing only the suffix and dropping the prefix. This has been centralized to only occur in `edit-profile.js`, where new profile destinations are saved and then passed around state with the prefix already attached.

**Method of Adding Translation Support**
The content in these menus was not being translated properly before because the destination name was being saved to state as a translated string; as a result, `react-18next` could not find it to translate it on language change. This PR fixes the problem by reconstructing the `destination` object type to store label (destination address), value (destination purpose), and position. This is then used the to construct and temporarily replace the label with the `<name> : <label>` string used as the menu options, which is then taken back apart upon being sent to state, rather than storing the translated label string. A similar tactic was taken with the network menu, where label is replaced temporarily by the translated string for populating the menu but restored to its former name on being saved to state via a new 3rd `networkName` property on the `network` object. I have added comments into the code to explain this process, as well.

This approach worked for everything except for profiles that had been made prior to this addition, as their destination purpose was saved without the `TripPurpose` prefix. This was fixed by adding in logic that checks for older profiles and adds the necessary prefix onto it for proper translation.


## Testing Instructions

 * Either run `./scripts/server` or `cd taui` → `yarn start`
 * Log in and add a couple of new destinations to your profile.
 * Since we are missing most translations for the 'network' menu, it may be helpful to add some kind of temporary and distinguishing marker to the relevant placeholder translation strings in the Spanish and Chinese translation files, e.g. I added ', es' and ', ch' to the end of the strings for the sake of testing.
 * Click on the translation buttons&mdash;you should see the value in both menus change. Select the menu and verify that all of the values in the menu have been translated, as well.
 * One of the issues I ran into was that while the default value and menu options were translating as they should, an option, once selected, would not translate on language change. Verify that this also works by selecting an option from the menu and then changing the language.
 * To check that the logic around old profiles works, checkout `develop` and add a new destination. Then return to this branch and check that your new destination translates as it should.


Resolves #391
